scimath/Mathematics/ClassicalStatistics.tcc

```diff
@@ -1125,69 +1127,114 @@ void ClassicalStatistics<CASA_STATP>::_createDataArray(
     vector<AccumType>& ary
 ) {
     _initIterators();
+    uInt nThreadsMax = _nThreadsMax();
+    PtrHolder<vector<AccumType> > tAry(
+        new vector<AccumType>[CACHE_PADDING*nThreadsMax], True
+    );
     while (True) {
         _initLoopVars();
-        if (_hasWeights) {
-            if (_hasMask) {
-                if (_hasRanges) {
-                    _populateArray(
-                        ary, _myData, _myWeights, _myCount,
-                        _myStride, _myMask, _maskStride, _myRanges, _myIsInclude
-                    );
-                }
-                else {
-                    _populateArray(
-                        ary, _myData, _myWeights,
-                        _myCount, _myStride, _myMask, _maskStride
-                    );
-                }
-            }
-            else if (_hasRanges) {
-                _populateArray(
-                    ary, _myData, _myWeights, _myCount,
-                    _myStride, _myRanges, _myIsInclude
-                );
-            }
-            else {
-                // has weights, but no mask nor ranges
-                _populateArray(
-                    ary, _myData, _myWeights, _myCount, _myStride
-                );
-            }
+        uInt nBlocks, nthreads;
+        uInt64 extra;
+        PtrHolder<DataIterator> dataIter;
+        PtrHolder<MaskIterator> maskIter;
+        PtrHolder<WeightsIterator> weightsIter;
+        PtrHolder<uInt64> offset;
+        _initThreadVars(
+            nBlocks, extra, nthreads, dataIter,
+            maskIter, weightsIter, offset, nThreadsMax
+        );
+#pragma omp parallel for num_threads(nthreads)
+        for (uInt i=0; i<nBlocks; ++i) {
+#ifdef _OPENMP
+            uInt tid = omp_get_thread_num();
+#else
+            uInt tid = 0;
+#endif
```

Comment:
this is repeated often enough now that it should go into its own function like nThreadsMax